File: alds1/7a_rooted_trees.c
#define dbg(fmt,...) fprintf(stderr,fmt,__VA_ARGS__)
#define dpri(x) dbg(#x ": %d\n", x)
#define dprs(x) dbg(#x ": %s\n", x)
#include <stdio.h>
#include <string.h>
typedef long long ll;
const int MYINF = 1e9+7;
typedef int bool;
const bool true = 1;
const bool false = 0;

typedef struct node_st {
	int parent;
	int child;
	int rsibling;
	int depth;
} node_t;


#define MAX_N 100010
int n;
node_t node[MAX_N];

const char *node_type_str(int id) {
	if (node[id].parent == -1) return "root";
	if (node[id].child == -1) return "leaf";
	else return "internal node";
}

void mark(int id, int d) {
	node[id].depth = d;
	if (node[id].child != -1) mark(node[id].child, d+1);
	if (node[id].rsibling != -1) mark(node[id].rsibling, d);
}

int main(int argc, char **argv)
{
	int i, j;
	memset(node, -1, sizeof(node_t) * MAX_N);
	scanf("%d", &n);
	for (i = 0; i < n; i++) {
		int id, k, left;
		scanf("%d %d", &id, &k);
		for (j = 0; j < k; j++) {
			if (j) {
				scanf("%d", &node[left].rsibling);
				node[node[left].rsibling].parent = id;
				left = node[left].rsibling;
			} else {
				scanf("%d", &node[id].child);
				node[node[id].child].parent = id;
				left = node[id].child;
			}
		}
	}

	for (i = 0; i < n; i++) {
		if (node[i].parent == -1) {
			mark(i, 0);
			break;
		}
	}

	for (i = 0; i < n; i++) {
		int child;
		const char *sp = "";

		printf("node %d: parent = %d, depth = %d, %s, [",
				i, node[i].parent, node[i].depth, node_type_str(i));
		child = node[i].child;
		while (child != -1) {
			printf("%s%d", sp, child);
			sp = ", ";
			child = node[child].rsibling;
		}
		printf("]\n");
	}
	return 0;
}

File: alds1/7a_rooted_trees.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

type Node struct {
	Parent int
	Child  []int
	Depth  int
}

func (n Node) Type() string {
	if n.Parent == -1 {
		return "root"
	} else if len(n.Child) == 0 {
		return "leaf"
	} else {
		return "internal node"
	}
}

func (n Node) SprintChild() string {
	s := []string{}
	for _, c := range n.Child {
		s = append(s, fmt.Sprintf("%d", c))
	}
	return strings.Join(s, ", ")
}

func markDepth(t []Node, id int, depth int) {
	t[id].Depth = depth
	for _, c := range t[id].Child {
		markDepth(t, c, depth+1)
	}
}

func main() {
	defer stdout.Flush()
	n := readInt()
	t := make([]Node, n)
	for i := 0; i < n; i++ {
		t[i].Parent = -1
	}
	for i := 0; i < n; i++ {
		id := readInt()
		d := readInt()
		t[id].Child = readIntSlice(d)
		for _, c := range t[id].Child {
			t[c].Parent = id
		}
	}
	p := -1
	for i := 0; i < n; i++ {
		if t[i].Parent == -1 {
			p = i
			break
		}
	}
	markDepth(t, p, 0)
	for i := 0; i < n; i++ {
		printf("node %d: parent = %d, depth = %d, %s, [%s]\n", i, t[i].Parent, t[i].Depth, t[i].Type(), t[i].SprintChild())
	}
}

var (
	stdin  *bufio.Reader
	stdout *bufio.Writer
)

func init() {
	stdin = bufio.NewReader(os.Stdin)
	stdout = bufio.NewWriter(os.Stdout)
}

func scanf(f string, args ...interface{}) (int, error) {
	n, err := fmt.Fscanf(stdin, f, args...)
	for {
		r, _, err := stdin.ReadRune()
		if err != nil || !unicode.IsSpace(r) {
			stdin.UnreadRune()
			break
		}
	}
	return n, err
}

func readInt() int {
	var n int
	scanf("%d", &n)
	return n
}

func readIntSlice(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		scanf("%d", &a[i])
	}
	return a
}

func readLengthAndSlice() []int {
	n := readInt()
	return readIntSlice(n)
}

func readString() string {
	var s string
	scanf("%s", &s)
	return s
}

func printf(f string, args ...interface{}) (int, error) {
	return fmt.Fprintf(stdout, f, args...)
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdout, args...)
}
